### PR TITLE
[RUNTIME] Introduce MetadataModule to separate code compilation/interpretation and weight initialization

### PR DESCRIPTION
**Pull request by zhiics**
This PR attempts to separate metadata initialization and source code compilation (when necessary). The discussion can be found here: https://discuss.tvm.ai/t/byoc-runtime-json-runtime-for-byoc/6579/15

The goal is to have a standalone wrapper (`MetadataModule`) to take care of module initialization for different runtimes. In the BYOC case, we save constants and code separately. The constant would be loaded by the wrapper while the code is handled by csourcemodule.

### CODE REVIEW

File: src/relay/backend/contrib/dnnl/codegen.cc [hunk 1]
```diff
@@ -392,22 +381,46 @@ class DNNLModuleCodegen : public CSourceModuleCodegenBase {
     code_stream_ << "using namespace tvm::runtime::contrib;\n";
     code_stream_ << "\n";
 
+    String func_symbol("all");
```
Comment (zhiics) [inline src/relay/backend/contrib/dnnl/codegen.cc:384]: I think it is okay that this is code a bit redundant to the one in c_codegen, because we are planning to remove this in the followup PRs and replace it with a json runtime.

**Comment by zhiics** [comment]
cc @tqchen @comaniac @junrushao1994 @FrozenGene @lhutton1 @trevor-m @masahi

File: python/tvm/target/source_module.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+# specific language governing permissions and limitations
+# under the License.
+# pylint: disable=no-else-return, unidiomatic-typecheck, undefined-variable, invalid-name, redefined-builtin
+"""
+Helper functions and classes for handling source and metdata.
+"""
+from tvm.runtime import _ffi_api
+
+class SourceMetadataModule:
+    """The module used to wrap both source and metadata."""
```
Comment (tqchen) [inline python/tvm/target/source_module.py:24]: Let us not expose it to the python side for now, as it is totally fine to get things from the mod itself. The only reason why have the graph runtime wrapper is that it is too commonly used

File: src/target/source/source_module.cc [hunk 3]
```diff
@@ -152,8 +153,92 @@ runtime::Module DeviceSourceModuleCreate(
...
+      return PackedFunc(
+          [sptr_to_self, this](TVMArgs args, TVMRetValue* rv) { *rv = this->source_type_; });
+    } else if (name == "get_symbol") {
+      return PackedFunc(
+          [sptr_to_self, this](TVMArgs args, TVMRetValue* rv) { *rv = this->func_symbol_; });
+    } else if (name == "get_vars") {
+      return PackedFunc(
+          [sptr_to_self, this](TVMArgs args, TVMRetValue* rv) { *rv = this->variables_; });
+    } else if (name == "get_metadata") {
+      return PackedFunc(
```
Comment (tqchen) [inline src/target/source/source_module.cc:198]: What is the relation between this class and the ModuleInitWrapper. It seems to be that at leastw can create a MetaDataModule without tying to CSourceModule(e.g. use it for LLVM target as well)

File: python/tvm/runtime/module.py [hunk 4]
```diff
@@ -33,6 +33,25 @@
 ProfileResult = namedtuple("ProfileResult", ["mean", "results"])
 
 
+def ModuleInitWrapper(variables, metadata):
+    """Create a module initialization wrapper.
```
[the remark this answers is not in the document]
Comment (zhiics) [reply python/tvm/runtime/module.py:37]: ah, I thought we want to allow users to play with these APIs in the frontend. NVM, I can hide them in the backend and some wrappers can be removed.

File: src/runtime/module_init_wrapper.cc [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+  }
+
+ private:
+  /*! \brief The stream to print constant data. */
+  std::ostringstream stream_;
+  /*! \brief variable name to NDArray mapping. */
+  StringNDArrayMap metadata_;
+};
+
+class ModuleInitWrapper : public runtime::ModuleNode {
```
[the remark this answers is not in the document]
Comment (zhiics) [reply src/runtime/module_init_wrapper.cc:98]: how about `MetadataInitModule` or just `MetadataModule`?

[the remark this answers is not in the document]
Comment (tqchen) [reply src/runtime/module_init_wrapper.cc:98]: MetadataModule sounds good

File: python/tvm/runtime/module.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (tqchen) [reply python/tvm/runtime/module.py]: Thanks! One thing that we want to be mindful of is to reduce the amount of concepts that an developer should learn. In most cases, we only want the developer to be aware of runtime.Module and not beyond that so the backend impl can change over time.

File: src/target/source/source_module.cc (hunk 3 above)
Comment (zhiics) [reply src/target/source/source_module.cc:218]: This module saves the data including source and metadata from the partitioned graphs. It is only used for packaging purpose. CSourceMoudle and later on other modules (e.g json runtime module) will take the code from it. The `ModuleInitWrapper` will take the variables and metadata from it. 

If we can let MetadataModule take all {var: ndarray} mapping and let CSourceModule get the needed variables. We should be able to remove this module as well.

File: src/runtime/module_init_wrapper.cc [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+namespace tvm {
+namespace runtime {
+
+using StringNDArrayMap = std::unordered_map<String, runtime::NDArray, ObjectHash, ObjectEqual>;
+
+class CSourceMetadataInitializer {
+ public:
+  explicit CSourceMetadataInitializer(const StringNDArrayMap& metadata) : metadata_(metadata) {}
+
+  template <typename T>
```
Comment (tqchen) [inline src/runtime/module_init_wrapper.cc:43]: Do we really need this one? My guess is that we can return MetaInitWrapper(that wraps CSourceModule) and our module serialization mechanism should take care of loading the meta-data back, without having to serialie the meta data to the C source.

Comment (tqchen) [reply src/runtime/module_init_wrapper.cc:43]: see also https://tvm.apache.org/docs/dev/introduction_to_module_serialization.html

File: src/target/source/source_module.cc (hunk 3 above)
Comment (tqchen) [reply src/target/source/source_module.cc:218]: It would be great to see if we can reduce to only MetaDataModule, and reuse mechanism here https://tvm.apache.org/docs/dev/introduction_to_module_serialization.html for serialization and packaging.

File: src/runtime/module_init_wrapper.cc (hunk 6 above)
Comment (zhiics) [reply src/runtime/module_init_wrapper.cc:43]: This is needed for partitioned subgraphs because they expect the plain c source data as the input.

Comment (tqchen) [reply src/runtime/module_init_wrapper.cc:43]: Can we change the subgraph generator to instead expose `InitMod0Func0(Array<NDArray>)` instead? so that we don't have to compile the meta data into the C source?

File: src/target/source/source_module.cc (hunk 3 above)
Comment (zhiics) [reply src/target/source/source_module.cc:218]: Okay, I left a TODO in the code for this. Let me give it a try in this PR.

File: src/runtime/module_init_wrapper.cc (hunk 6 above)
Comment (zhiics) [reply src/runtime/module_init_wrapper.cc:43]: This should be possible. Let me spend some time to refactor the c source code generator over the weekend.

File: src/target/source/source_module.cc (hunk 3 above)
Comment (zhiics) [reply src/target/source/source_module.cc:218]: @tqchen I refactored CSourceModule generator for subfunctions to initialize the submodules. Now I am working on this. But I have a few questions

Suppose we have:
```
MetadataModule:
    - Graphruntime LLVM module(uses `p0 = const0, p1 = const1`)
    - Accelerator CSourceModule0, foo_0 (uses `a = const1,  b = const2`)
    - Accelerator CSourceModule1, foo_1 (uses `c = const1, d = const3`)
```

- How do we collect all required Metadata? Maybe we can collect them through graph_runtime_codegen/vmcompiler because the params (var: ndarray pairs) used by graphruntime/VM is from there. We can visit into other functions to collect for the other two submodules. 
- How do we dispatch the correct constants to the corresponding submodule,  `InitFoo0(const1, const2)`? It looks to me that MetadataModule is not aware of any const to submodule mapping. I current stored all the variables and ndarrays. Should we also need to have each submodule store the required const_vars? By this, `MetadataModule` can query which ndarrays a submodule would require by looking at the list of `const_vars`. However, this would mean we need to have a `const_vars/params` field for different runtime modules.

Am I missing thing?

File: python/tvm/runtime/module.py [hunk 7]
```diff
@@ -222,21 +222,29 @@ def evaluator(*args):
...
         # append root module
         visited.add(self)
         stack.append(self)
         while stack:
             module = stack.pop()
             if module._dso_exportable():
                 dso_modules.append(module)
+            elif module.type_key == "module_init":
+                assert not metadata_init, \
+                        "At most one module initializer is allowed"
```
[the remark this answers is not in the document]
Comment (zhiics) [reply python/tvm/runtime/module.py:240]: The python side changes will be all removed.

[the remark this answers is not in the document]
Comment (lhutton1) [reply python/tvm/runtime/module.py:240]: Makes sense thanks, I was thinking one MetadataModule would be created for each backend.

**Comment by zhiics** [comment]
@tqchen I removed all wrappers, made `MetadataModule` internal, and dispatched the initialization of each submodule per function/symbol base. There is now no user API change. But I had to add a field the CSourceModule to let it know what metadata it needs. This, however, doesn't affect the compilation and the original use/semantic of `CSourceModule` because we only need this information to help build up the `MetadataModule`.

**Comment by tqchen** [comment]
cc @mbaret @FrozenGene @comaniac it would be great if you can help to take a look as well

File: python/tvm/contrib/graph_runtime.py [hunk 8]
```diff
@@ -162,7 +162,11 @@ def set_input(self, key=None, value=None, **params):
             keys = list(params.keys())
             keys.sort(key=lambda x: -np.prod(params[x].shape))
             for k in keys:
-                self._get_input(k).copyfrom(params[k])
+                # TODO(zhiics) Skip the weights for submodule in a better way.
+                # We could get all inputs required by graphruntime first,
+                # we should use MetadataModule for initialization.
+                if "_const_" not in k:
```
Comment (FrozenGene) [inline python/tvm/contrib/graph_runtime.py:168]: Is `_const_` only existed for external module?

**Comment by FrozenGene** [comment]
Maybe I miss something. so maybe this is considered. That is I don't see `unwrap_modules`. I ask this because I think it is related with our next module based interface runtime. The usage like [this](https://github.com/apache/incubator-tvm/pull/5753/files#diff-82bcc5ddf3cb06075848655807dfcbd5R136-R157):
```python
with relay.build_config(opt_level=3):
    complied_graph_lib = relay.build_module.build(
        mod, "llvm", params=params)

from tvm.contrib import util
temp = util.tempdir()
file_name = "deploy_lib.so"
path_lib = temp.relpath(file_name)
complied_graph_lib.export_library(path_lib)
loaded_lib = tvm.runtime.load_module(path_lib)
ctx = tvm.cpu(0)
gmod = loaded_lib['default'](ctx)
set_input = gmod["set_input"]
run = gmod["run"]
get_output = gmod["get_output"]
data = np.random.uniform(-1, 1, size=(1, 3, 224, 224)).astype("float32")
set_input("data", tvm.nd.array(data))
run()
out = get_output(0).asnumpy()
```
As you could see we will wrap `params` / `lib` / `json` into one lib. So I want to see the logic of `unwrap_modules` and to see how to connect with it. 

THANKS!

File: python/tvm/contrib/graph_runtime.py (hunk 8 above)
Comment (zhiics) [reply python/tvm/contrib/graph_runtime.py:168]: Yes, the constants for graph runtime all start with "p"

https://github.com/apache/incubator-tvm/blob/3e72be58f362a21dbcc1de36f9dbed216e854baf/src/relay/backend/graph_runtime_codegen.cc#L314

However, as the TODO says, we should have a better way to handle this. But graph runtime current doesn't really distinguish input/weights. I think we can use this simple workaround from the frontend for now.

**Comment by zhiics** [comment]
@FrozenGene The current implementation doesn't actually change any user/frontend APIs. We added the `MetadataModule` to take care of weight initialization and separate it from code compilation. So you would have:

```
MetadataModule:
    - DSO module
    - Accelerator CSourceModule0
    - Accelerator CSourceModule1
```

As you can see, `MetadataModule` manages the params, but we will still have code in the submodules. The serialization and deserialization make sure that the module structure is always maintained correctly. Therefore, we can still have params/lib/json in one module. 

In other words, it should not affect the way we package the final artifacts, but it changes the way we compile certain modules (e.g. CSourceModule) and the way we initialize modules that need constant data.

File: src/relay/backend/contrib/codegen_c/codegen.cc [hunk 9]
```diff
@@ -76,43 +77,31 @@ class CodegenC : public MemoizedExprTranslator<std::vector<Output>>, public Code
...
-    output.name = "const_" + std::to_string(const_idx_++);
-
-    runtime::NDArray array = cn->data;
-    const auto& shape = array.Shape();
-
-    // Get the number of elements.
-    int64_t num_elems = 1;
-    for (auto i : shape) num_elems *= i;
-
+    // Get const: static_cast<float*>(dnnl_0_consts[0]->data)
```
Comment (comaniac) [inline src/relay/backend/contrib/codegen_c/codegen.cc:84]: s/dnnl_0/gcc_0/

File: src/relay/backend/contrib/codegen_c/codegen_c.h [hunk 10]
```diff
@@ -190,7 +209,12 @@ class CodegenCBase {
    */
   std::string JitImpl(const std::string& ext_func_id, const Array<Var>& args,
                       const std::vector<std::string>& buf_decl,
-                      const std::vector<std::string>& body, const std::vector<Output>& outs) {
+                      const std::vector<std::string>& body, const std::string& const_arr,
```
Comment (comaniac) [inline src/relay/backend/contrib/codegen_c/codegen_c.h:212]: `const_arr` seems a bit confusion. Maybe `const_arr_name` or something like that. (ditto in some other places)

File: src/relay/backend/contrib/codegen_c/codegen.cc [hunk 11]
```diff
@@ -76,43 +77,31 @@ class CodegenC : public MemoizedExprTranslator<std::vector<Output>>, public Code
...
+                 << "_consts.empty()) << \"C source module hasn't been initialized.\";\n";
+      ext_func_body_.insert(ext_func_body_.begin(), buf_stream.str());
     }
-    buf_stream << "};";
-    ext_func_body.insert(ext_func_body.begin(), buf_stream.str());
+
+    CHECK(dtype == "float" || dtype == "int") << "Only float and int are supported for now.";
+    output.dtype = dtype;
+
+    std::string const_var_name = ext_func_id_ + "_const_" + std::to_string(const_idx_++);
```
Comment (comaniac) [inline src/relay/backend/contrib/codegen_c/codegen.cc:103]: As `_const_` is now a keyword for runtime to take actions, we may generate `const_var_name` in the base class to make sure users won't violate miss it.

File: src/runtime/metadata_module.cc [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+      // Get the initialization function from the imported modules.
+      std::string init_name = "__init_" + symbol;
+      init = it.GetFunction(init_name, false);
+      if (init != nullptr) {
+        auto md = GetRequiredMetadata(symbol);
+        // Initialize the module with metadata.
+        init(md);
+        break;
+      }
+    }
```
Comment (comaniac) [inline src/runtime/metadata_module.cc:116]: Better to have a CHECK to make sure it was initialized correctly.

File: src/runtime/metadata_module.cc [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+  const char* type_key() const { return "metadata"; }
+
+  /*!
+   * \brief Get the list of metadata that is required by the given module.
+   * \param symbol The symbol that is being queried.
+   * \return The list of needed NDArray.
+   */
+  Array<NDArray> GetRequiredMetadata(const std::string& symbol) {
+    Array<NDArray> ret;
+    CHECK_GT(sym_vars_.count(symbol), 0U) << "Not symbol is recorded for " << symbol;
```
Comment (comaniac) [inline src/runtime/metadata_module.cc:82]: s/Not/No/

File: src/runtime/metadata_module.cc [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+    auto n = make_object<MetadataModuleNode>(metadata, sym_vars);
+    return Module(n);
+  }
+
+ private:
+  /*!
+   * \brief Record if a module is initialized. It is needed by imported
+   * modules using execution engine.
+   */
+  std::unordered_map<std::string, bool> initialized_;
```
Comment (comaniac) [inline src/runtime/metadata_module.cc:203]: `unordered_set` fits better for this case, but I suppose `unordered_map` will be replaced with `runtime::Map` in the future so it should be fine to keep it.

File: src/runtime/metadata_module.cc [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+  }
+
+  static Module LoadFromBinary(void* strm) {
+    dmlc::Stream* stream = static_cast<dmlc::Stream*>(strm);
+
+    // Load the variables.
+    std::vector<std::string> variables;
+    CHECK(stream->Read(&variables)) << "Loading variables failed";
+    uint64_t sz;
+    CHECK(stream->Read(&sz, sizeof(sz))) << "Loading medata size failed";
```
Comment (comaniac) [inline src/runtime/metadata_module.cc:160]: s/medata/metadata/

File: src/runtime/metadata_module.cc [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+    CHECK_EQ(static_cast<size_t>(sz), variables.size())
```
Comment (comaniac) [inline src/runtime/metadata_module.cc:161]: I think we can keep this cast to make `size_t m_size = static_cast<size_t>(sz)` so that the rest loops in this function can all use `size_t i`.

File: src/relay/backend/contrib/codegen_c/codegen_c.h [hunk 17]
```diff
@@ -299,6 +301,55 @@ class CodegenCBase {
...
+   * \brief Creates a checker to check if the NDArray pool is initialized
+   *
+   * \param symobl The Symbol of the current function
+   *
+   * \return The created checker
+   */
+  std::string CreateInitChecker(const std::string& symbol) const {
+    std::ostringstream oss;
+    oss << "CHECK(!" << symbol
+        << "_consts.empty()) << \"DNNL source module hasn't been initialized.\";\n";
```
Comment (comaniac) [inline src/relay/backend/contrib/codegen_c/codegen_c.h:314]: s/DNNL/C/

File: python/tvm/contrib/graph_runtime.py [hunk 18]
```diff
@@ -162,7 +162,11 @@ def set_input(self, key=None, value=None, **params):
             keys = list(params.keys())
             keys.sort(key=lambda x: -np.prod(params[x].shape))
             for k in keys:
-                self._get_input(k).copyfrom(params[k])
+                # TODO(zhiics) Skip the weights for submodule in a better way.
+                # We could get all inputs required by graphruntime first,
+                # we should use MetadataModule for initialization.
```
Comment (tqchen) [inline python/tvm/contrib/graph_runtime.py:167]: This part of the logic seems to be a bit too hacky, cryptic Shall we simply skip inside the set_params instead? If the weights are emedded in the metadata, do we still allow overriding of weights?

Comment (zhiics) [reply python/tvm/contrib/graph_runtime.py:167]: Yeah, we can do some checking at the C++ side as well. We will need to change graph runtime later to let it have a "__init" function so that we can initialize it with params. And we will need to remove some of the current ways to set the params up. For example, the `set_input`  API should not take `**params` and the chunk of hacky code will be removed. That needs some more thinking and may worth an RFC.

Comment (tqchen) [reply python/tvm/contrib/graph_runtime.py]: Perhaps it is possible to automatically call the `__init` function in module construction and binary loading time ?

Comment (zhiics) [reply python/tvm/contrib/graph_runtime.py]: yeah, that is something what I was thinking of as well. Let me spend some time to think about some possible ways. I will post an RFC with some plans because graph runtime affects apparently every downstream users.

File: python/tvm/contrib/graph_runtime.py [hunk 19]
```diff
@@ -162,7 +162,12 @@ def set_input(self, key=None, value=None, **params):
             keys = list(params.keys())
             keys.sort(key=lambda x: -np.prod(params[x].shape))
             for k in keys:
-                self._get_input(k).copyfrom(params[k])
+                # TODO(zhiics) Skip the weights for submodule in a better way.
+                # We should use MetadataModule for initialization and remove
```
Comment (tqchen) [inline python/tvm/contrib/graph_runtime.py:166]: Shall we shoot an warning at least if val is not available?

Comment (zhiics) [reply python/tvm/contrib/graph_runtime.py:166]: Yes, there is a warning in the C++ side.

https://github.com/apache/incubator-tvm/blob/d8c80c382f02052b07da1235190a5b6c7acea994/src/runtime/graph/graph_runtime.cc#L93
